server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotTaskRunnerTests.java

```diff
@@ -151,43 +158,78 @@ public void testShardSnapshotTaskRunner() throws Exception {
-        SnapshotShardContext s3Context = dummyContext(s3, s2StartTime);
-        // Shard snapshot and file snapshot tasks for earlier snapshots have higher priority
-        assertThat(
-            workers.new ShardSnapshotTask(s1Context).compareTo(
-                workers.new FileSnapshotTask(s2Context, ShardSnapshotTaskRunnerTests::dummyFileInfo, listener)
-            ),
-            lessThan(0)
-        );
-        assertThat(
-            workers.new FileSnapshotTask(s1Context, ShardSnapshotTaskRunnerTests::dummyFileInfo, listener).compareTo(
-                workers.new ShardSnapshotTask(s2Context)
-            ),
-            lessThan(0)
-        );
-        // Two tasks with the same start time and of the same type are ordered by snapshot UUID
-        assertThat(workers.new ShardSnapshotTask(s2Context).compareTo(workers.new ShardSnapshotTask(s3Context)), lessThan(0));
-        assertThat(
-            workers.new FileSnapshotTask(s2Context, ShardSnapshotTaskRunnerTests::dummyFileInfo, listener).compareTo(
-                workers.new ShardSnapshotTask(s3Context)
-            ),
-            lessThan(0)
-        );
-        // Shard snapshot task has a higher priority over file snapshot within the same snapshot
-        assertThat(
-            workers.new ShardSnapshotTask(s1Context).compareTo(
-                workers.new FileSnapshotTask(s1Context, ShardSnapshotTaskRunnerTests::dummyFileInfo, listener)
-            ),
-            lessThan(0)
-        );
+
+        record CapturedTask(SnapshotShardContext context, @Nullable BlobStoreIndexShardSnapshot.FileInfo fileInfo) {
+            CapturedTask(SnapshotShardContext context) {
+                this(context, null);
+            }
+        }
+
+        final List<CapturedTask> tasksInExpectedOrder = new ArrayList<>();
+
+        // first snapshot, one shard, one file, but should execute the shard-level task before the file task
+        final var earlyStartTime = randomLongBetween(1L, Long.MAX_VALUE - 1000);
+        final var s1Context = dummyContext(new SnapshotId(randomIdentifier(), randomUUID()), earlyStartTime);
+        tasksInExpectedOrder.add(new CapturedTask(s1Context));
+        tasksInExpectedOrder.add(new CapturedTask(s1Context, dummyFileInfo()));
+
+        // second snapshot, also one shard and one file, starts later than the first
+        final var laterStartTime = randomLongBetween(earlyStartTime + 1, Long.MAX_VALUE);
+        final var s2Context = dummyContext(new SnapshotId(randomIdentifier(), "early-uuid"), laterStartTime);
+        tasksInExpectedOrder.add(new CapturedTask(s2Context));
+        tasksInExpectedOrder.add(new CapturedTask(s2Context, dummyFileInfo()));
+
+        // third snapshot, starts at the same time as the second but has a later UUID
+        final var snapshotId3 = new SnapshotId(randomIdentifier(), "later-uuid");
+
+        // the third snapshot has three shards, and their respective tasks should execute in shard-id then index-name order:
```

Comment:
Where does the index-name order come in? I don't see it in COMPARATOR 🤔

Reply:
Here:

https://github.com/elastic/elasticsearch/blob/c16f2c9f877de8d31afc8cb9936d2b249b4be63f/server/src/main/java/org/elasticsearch/index/shard/ShardId.java#L117

(in `ShardId#compareTo`)